### ISSUE TITLE
Time zones don't support fold

### ISSUE THREAD

**Issue by pganssle** [mahmoud/boltons#141]
The `LocalTZInfo` and `USTimeZone` functions do not support the PEP495 `fold` attribute, and will thus be wrong around ambiguous datetimes. See:

```
from dateutil import tz
from boltons.timeutils import LocalTZInfo, Eastern

from datetime import datetime

def fmt_offset(dt):
    # Default timedelta formatting is ugly
    td = dt.utcoffset()
    fmt_str = ('-' if td < timedelta(0) else '') + '{}'
    return fmt_str.format(abs(td))

EASTERN_DU = tz.gettz('US/Eastern')
LOCAL_DU = tz.tzlocal()

EASTERN_BO = Eastern
LOCAL_BO = LocalTZInfo()

dt_unfolded = datetime(2011, 11, 6, 1, 30)
dt_folded = tz.enfold(dt_unfolded, fold=1)

for tz_du, tz_bo in ((EASTERN_DU, EASTERN_BO), (LOCAL_DU, LOCAL_BO)):
    dt_du_u = dt_unfolded.replace(tzinfo=tz_du)
    dt_bo_u = dt_unfolded.replace(tzinfo=tz_bo)

    dt_du_f = dt_folded.replace(tzinfo=tz_du)
    dt_bo_f = dt_folded.replace(tzinfo=tz_bo)

    print("(Dateutil, folded)   Offset: {} TZName: {}".format(fmt_offset(dt_du_f),
                                                              dt_du_f.tzname()))
    print("(boltons, folded)    Offset: {} TZName: {}".format(fmt_offset(dt_bo_f),
                                                             dt_bo_f.tzname()))
    print()

    print("(Dateutil, unfolded) Offset: {} TZName: {}".format(fmt_offset(dt_du_u),
                                                                dt_du_u.tzname()))
    print("(boltons, unfolded)  Offset: {} TZName: {}".format(fmt_offset(dt_bo_u),
                                                               dt_bo_u.tzname()))
    print()

###
### Result:
"""
(Dateutil, folded)   Offset: -5:00:00 TZName: EST
(boltons, folded)    Offset: -5:00:00 TZName: EST

(Dateutil, unfolded) Offset: -4:00:00 TZName: EDT
(boltons, unfolded)  Offset: -5:00:00 TZName: EST

(Dateutil, folded)   Offset: -5:00:00 TZName: EST
(boltons, folded)    Offset: -4:00:00 TZName: EDT

(Dateutil, unfolded) Offset: -4:00:00 TZName: EDT
(boltons, unfolded)  Offset: -4:00:00 TZName: EDT
"""
```

It believe it's slightly easier to fix this if you only support the built-in fold attribute in Python 3.6, otherwise you need to add a backport `_DatetimeWithFold` object. 

[Here](https://github.com/dateutil/dateutil/blob/master/dateutil/tz/_common.py#L31) is how that is handled in `dateutil`, [here](https://github.com/dateutil/dateutil/blob/master/dateutil/tz/tz.py#L159) is dateutil's implementation of `tzlocal` (some of the fold handling happens in the parent class). I don't have a ready-made `USTimeZone` implementation, but I believe the example in the Python documentation was updated accordingly.

Note that `dateutil` is BSD-licensed, so taking code from there should be compatible with this project.

**Comment by pganssle** [comment]
Interesting that the `LocalTZInfo` and `USTimezone` implementations seem to have opposite defaults for ambiguous datetimes.

**Comment by mahmoud** [comment]
Very interesting all around! But, hmm, were the Python docs updated, actually? Look at the code examples [below here](https://docs.python.org/3.7/library/datetime.html#datetime.tzinfo.fromutc).

The docs say: 

> Applications that can’t bear wall-time ambiguities should explicitly check the value of the fold attribute or avoid using hybrid tzinfo subclasses; there are no ambiguities when using timezone, or any other fixed-offset tzinfo subclass (such as a class representing only EST (fixed offset -5 hours), or only EDT (fixed offset -4 hours)).

Seems to suggest they don't think they need fold?

**Comment by pganssle** [comment]
@mahmoud Yes, the code examples were updated. See this part in the `LocalTimezone` example:

```python
        # Detect fold
        fold = (args == _time.localtime(stamp - dst_diff))
        return datetime(*args, microsecond=dt.microsecond,
                        tzinfo=self, fold=fold)
```

`fold` was not a keyword argument before Python 3.6.

> Seems to suggest they don't think they need fold?

Completely unambiguous time zones (zones with no DST ever) don't need to handle fold because no folds occur in those zones. If you just need a "point in time" you can also use `timezone` or `dateutil.tz.tzoffset()` (I think there's a bolton called `FixedOffset` maybe?), because that is a single fixed offset with no knowledge of DST - essentially equivalent to UTC.

I think what they are basically saying is that if you have an application where being wrong about the time during DST transitions is not a big issue, you don't have to bother with handling the fold. As a library developer, though, it's probably best to have your library do the same thing, since most people won't realize the library they are using has an edge case until they run into it.

I'd say the simplest solution is to just opportunistically use/set the `fold` information - if you provide a `datetime` object that has a `fold` attribute, use it, otherwise assume it is 0. Users can upgrade to Python 3.6+ if they want the correct behavior, or can select a library like `dateutil` (or, for a more all-inclusive solution, something like `pendulum`), which includes a backport.

If you want the behavior to be consistent across versions, you can always just import the `dateutil` backport solution (which is in turn based on Alexander Beloposky's MIT-licensed backport).
